### PR TITLE
WIP: Build presto-cli using GraalVM native-image

### PR DESCRIPTION
**Pull request by wendigo**

### CODE REVIEW

File: presto-cli/pom.xml [hunk 1]
```diff
@@ -88,7 +89,7 @@
 
         <dependency>
             <groupId>org.jline</groupId>
-            <artifactId>jline-terminal-jna</artifactId>
+            <artifactId>jline-terminal-jansi</artifactId>
```
Comment (electrum) [inline presto-cli/pom.xml]: We want to move away from Jansi. See #2878

File: presto-cli/pom.xml [hunk 2]
```diff
@@ -120,13 +121,71 @@
             <scope>test</scope>
         </dependency>
 
+        <dependency>
```
Comment (electrum) [inline presto-cli/pom.xml]: Put these in a different section, not in the middle of the test dependencies

**Comment by electrum** [comment]
Where do all these resource files come from? Are they auto generated?

File: presto-cli/pom.xml (hunk 1 above)
Comment (wendigo) [reply presto-cli/pom.xml]: I couldn't get JNA working on native-image (that's why I've switched to JANSI)

**Comment by wendigo** [comment]
@electrum some yes (using agent), some by hand (agent cannot pick everything)

**Comment by wendigo** [comment]
@electrum I've managed to go back to jna providing proper jna config by studying https://github.com/java-native-access/jna/blob/master/native/dispatch.c

File: presto-cli/pom.xml (hunk 1 above)
Comment (wendigo) [reply presto-cli/pom.xml]: Solved.

**Comment by wendigo** [comment]
```
time presto-cli/target/presto-cli --execute 'select count(*) from tpch.sf1.lineitem'
"6001215"
presto-cli/target/presto-cli --execute   0.03s user 0.03s system 4% cpu 1.357 total
```

vs

```
time presto-cli/target/presto-cli-331-SNAPSHOT-executable.jar --execute 'select count(*) from tpch.sf1.lineitem'
"6001215"
presto-cli/target/presto-cli-331-SNAPSHOT-executable.jar --execute   1.71s user 0.19s system 91% cpu 2.087 total
```

File: presto-cli/src/main/resources/META-INF/native-image/resource-config.json [hunk 3]
```diff
@@ -0,0 +1,29 @@
+{
+  "resources":[
+    {"pattern":"\\QMETA-INF/services/org.jline.terminal.spi.JnaSupport\\E"},
+    {"pattern":"\\Qcom/sun/jna/darwin/libjnidispatch.jnilib\\E"},
+    {"pattern":"\\Qorg/jline/utils/capabilities.txt\\E"},
+    {"pattern":"\\Qorg/jline/utils/xterm-256color.caps\\E"},
+    {"pattern": "org/jline/utils/.*caps$"},
+    {"pattern": "org/jline/utils/capabilities\\.txt$"},
```
Comment (electrum) [inline presto-cli/src/main/resources/META-INF/native-image/resource-config.json]: These look wrong. We don’t use AWS or Jansi.

File: presto-cli/src/main/resources/META-INF/native-image/resource-config.json [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    {"pattern": "com/amazonaws/global/handlers/request\\.handler2s"},
+      {"pattern": ".*properties$"},
+      {"pattern": "META-INF/native/windows64/jansi.dll"},
+      {"pattern":"META-INF/native/osx/libjansi.jnilib"},
+      {"pattern":"org/fusesource/jansi/META-INF/native/osx/libjansi.jnilib"},
+      {"pattern":"org/fusesource/jansi/jansi/META-INF/native/osx/libjansi.jnilib"},
+      {"pattern":"org/fusesource/jansi/jansi/jansi/META-INF/native/osx/libjansi.jnilib"},
+      {"pattern":".*/libjansi.jnilib$"},
+      {"pattern":"META-INF/native/osx/libjansi.jnilib"},
+      {"pattern":"com/facebook/presto/spi/type/zone-index.properties"},
```
Comment (electrum) [inline presto-cli/src/main/resources/META-INF/native-image/resource-config.json]: Or this

File: presto-cli/src/main/resources/META-INF/native-image/resource-config.json [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+      {"pattern":"jline/console/completer/CandidateListCompletionHandler.properties"},
+      {"pattern":"META-INF/services/org.jline.terminal.spi.JansiSupport"},
+      {"pattern":"META-INF/services/org.jline.terminal.spi.JnaSupport"},
+      {"pattern":"logback.xml"},
```
Comment (electrum) [inline presto-cli/src/main/resources/META-INF/native-image/resource-config.json]: Where does this come from?

File: presto-cli/src/main/resources/META-INF/native-image/resource-config.json (hunk 3 above)
Comment (wendigo) [reply presto-cli/src/main/resources/META-INF/native-image/resource-config.json]: this was autocollected by agent, probably needs cleanup after I switched back to jline-jna from jline-jansi

File: presto-cli/src/main/resources/META-INF/native-image/resource-config.json (hunk 5 above)
Comment (wendigo) [reply presto-cli/src/main/resources/META-INF/native-image/resource-config.json]: I'll cleanup configs when I have working configuration on linux and windows as well

**Comment by electrum** [comment]
Is there any performance improvement when using the CLI to download data? The example you showed could be mostly due to startup overhead.

**Comment by wendigo** [comment]
https://github.com/oracle/graal/issues/2251

File: presto-cli/src/main/resources/META-INF/native-image/jni-config.json [hunk 6]
```diff
@@ -0,0 +1,173 @@
+[
```
Comment (kokosing) [inline presto-cli/src/main/resources/META-INF/native-image/jni-config.json]: How did you generate this? Can you please add step to github actions that tries to generate them and see if there are no changes comparing to what was commited? To make sure this file is up to date. Something like: https://github.com/prestosql/presto/blob/master/presto-main/bin/check_webui.sh

Comment (electrum) [reply presto-cli/src/main/resources/META-INF/native-image/jni-config.json]: If we can generate them, we should do that at build time and not check it in at all.

**Pull request closed without merging**